include/isotp.h

```diff
@@ -0,0 +1,350 @@
+ * @param rx_addr     Identifier for FC frames.
+ * @param tx_addr     Identifier for outgoing frames the receiver listens on.
+ * @param complete_cb Function called on completion or NULL.
+ * @param cb_arg      Argument passed to the complete callback.
+ * @param timeout     Timeout for buffer allocation.
+ *
+ * @retval ISO_TP_N_OK on success or any of ISO_TP_N_* on error
+ */
+int iso_tp_send_ctx_buf(struct device *can_dev,
+			const u8_t *data, size_t len,
+			const struct iso_tp_msg_id *tx_addr,
+			const struct iso_tp_msg_id *rx_addr,
+			isotp_tx_callback_t complete_cb, void *cb_arg,
+			s32_t timeout);
+
+/**
+ * @brief Send data with buffered context
+ *
+ * This function is similar to iso_tp_send_ctx_buf but the data is carried in
+ * a net_buf. net_buf_unref is called on the net_buf when sending is completed.
+ *
+ * @param can_dev     The CAN device to be used for sending and receiving.
+ * @param data        Data to be sent.
+ * @param len         Length of the data to be sent.
+ * @param rx_addr     Identifier for FC frames.
+ * @param tx_addr     Identifier for outgoing frames the receiver listens on.
+ * @param complete_cb Function called on completion or NULL.
+ * @param cb_arg      Argument passed to the complete callback.
+ * @param timeout     Timeout for buffer allocation.
+ *
+ * @retval ISO_TP_N_OK on success or any of ISO_TP_* on error
+ */
+int iso_tp_send_net_ctx_buf(struct device *can_dev,
+			    struct net_buf *data,
+			    const struct iso_tp_msg_id *tx_addr,
+			    const struct iso_tp_msg_id *rx_addr,
+			    isotp_tx_callback_t complete_cb, void *cb_arg,
+			    s32_t timeout);
+
+#endif /*CONFIG_CAN_ISOTP_ENABLE_CONTEXT_BUFFERS*/
+
+#if defined(CONFIG_CAN_ISOTP_USE_TX_BUF) && \
+	defined(CONFIG_CAN_ISOTP_ENABLE_CONTEXT_BUFFERS)
+/**
+ * @brief Send data with buffered context
+ *
+ * This function is similar to iso_tp_send but the context is automatically
+ * allocated from an internal pool and the data to be send is buffered in an
+ * internal net_buff.
+ *
+ * @param can_dev     The CAN device to be used for sending and receiving.
+ * @param data        Data to be sent.
+ * @param len         Length of the data to be sent.
+ * @param rx_addr     Identifier for FC frames.
+ * @param tx_addr     Identifier for outgoing frames the receiver listens on.
+ * @param complete_cb Function called on completion or NULL.
+ * @param cb_arg      Argument passed to the complete callback.
+ * @param timeout     Timeout for buffer allocation.
+ *
+ * @retval ISO_TP_N_OK on success or any of ISO_TP_* on error
+ */
+int iso_tp_send_buf(struct device *can_dev,
+		    const u8_t *data, size_t len,
+		    const struct iso_tp_msg_id *tx_addr,
+		    const struct iso_tp_msg_id *rx_addr,
+		    isotp_tx_callback_t complete_cb, void *cb_arg,
+		    s32_t timeout);
+#endif
+
+/** @cond INTERNAL_HIDDEN */
+
+struct iso_tp_callback {
+	isotp_tx_callback_t cb;
+	void *arg;
+};
+
+struct iso_tp_send_ctx {
```

Comment:
After a quick look, the send and recv structs look like there would be holes in them. Could you check with pahole program that the amount of memory occupied by these structs is the minimal one?

Reply:
Sure. The biggest hole is the `struct iso_tp_msg_id`. Ic ould use `__packed` to save two bytes. It would save four bytes per context. Do you think it is a good idea?

Reply:
The __packed should only be used if we are for example casting to network data that we have received. So do not add that attribute here.